src/box/lua/load_cfg.lua

```diff
@@ -1011,13 +1013,45 @@ local function load_cfg(cfg)
             box[k] = v
         end
     end
+
     setmetatable(box, nil)
-    box.cfg = setmetatable(cfg,
+    -- There is only a single way for setting a value to a box option -
+    -- call 'box.cfg{ option_name=option_value }'. Direct modifications
+    -- of table box.cfg are not allowed. On attempt to make a direct
+    -- modifications (setting a new option or updating an existing one) we raise
+    -- an error. For implementation this we are using metamethods __newindex
+    -- and __index. Both __index and __newindex are relevant only when the
+    -- index does not exist in the table. The only way to catch all accesses to
+    -- a table is to keep it empty. So we use an empty proxy table and proxying
+    -- setting and reading to another table.
+    raw_cfg = table.copy(cfg)
+    local proxy_table = {}
+    box.cfg = setmetatable(proxy_table,
         {
-            __newindex = function(table, index) -- luacheck: no unused args
+            -- tarantoolctl (extra/dist/tarantoolctl.in) expects a metatable
+            -- with metamethods, not a string. Thus we couldn't use __metatable
+            -- instead.
+            __name = 'box_cfg',
+            __index = raw_cfg,
+            __newindex = function(self, key, value) -- luacheck: no unused args
+                if template_cfg[key] ~= nil then
+                    local v = '<...>'
+                    if type(value) ~= 'table' then
```

Comment:
Minor: The only thing that bothers me here is the condition.

Is it possible to pass other "compex" types (such as cdata/udata/function) as a `box.cfg` parameter value? If not, then everything is fine; if there are doubts we can just leave it as is and fix if/when one is found: yeah, it's bad when there is garbage in the error message, but not critical.

Feel free to ignore.